pdl_syntax.md

```diff
@@ -0,0 +1,843 @@
+}
+```
+
+
+## Map Type
+
+Pegasus maps are defined with a values type and an optional key type.
+
+For example: 
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+record WithPrimitivesMap {
+  ints: map[string, int]
+  longs: map[string, long]
+  floats: map[string, float]
+  doubles: map[string, double]
+  booleans: map[string, boolean]
+  strings: map[string, string]
+  bytes: map[string, bytes]
+}
+```
+
+Primitive maps with default values: 
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+record WithPrimitivesMapDefaults {
+  ints: map[string, int] = { "int1": 1, "int2": 2, "int3": 3 }
+  longs: map[string, long] = { "long1": 3000000000, "long2": 4000000000 }
+  floats: map[string, float] = { "float1": 3.3, "float2": 2.1 }
+  doubles: map[string, double] = {"double1": 4.4E38, "double2": 3.1E24}
+  booleans: map[string, boolean] = { "boolean1": true, "boolean2": true, "boolean3": false }
+  strings: map[string, string] = { "string1": "hello", "string2": "world" }
+  bytes: map[string, bytes] = { "bytes": "\u0007" }
+}
+```
+
+Complex Types Map:
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+import com.linkedin.pegasus.generator.test.idl.enums.Fruits
+import com.linkedin.pegasus.generator.test.idl.records.Empty
+import com.linkedin.pegasus.generator.test.idl.records.Simple
+import com.linkedin.pegasus.generator.test.idl.`fixed`.Fixed8
+
+record WithComplexTypesMap {
+  empties: map[string, Empty]
+  fruits: map[string, Fruits]
+  arrays: map[string, array[Simple]]
+  maps: map[string, map[string, Simple]]
+  unions: map[
+    string,
+    typeref WithComplexTypesMapUnion = union[int, string, Simple]
+  ]
+  `fixed`: map[string, Fixed8]
+}
+```
+
+**Note**: 
+The key must always be "string".
+
+
+## Union Type
+
+A union type may be defined with any number of member types. Member type can be primitive, record, enum, map or array. Unions are not allowed as members inside an union.
+
+For example:
+```
+namespace com.linkedin.pegasus.generator.test.idl.unions
+
+record WithPrimitivesUnion {
+  value: union[int, long, float, double, boolean, string, bytes]
+}
+```
+
+The member type names also serve as the “member keys” (also called as “union tags”), and identify which union member type data holds.
+To define a field of a record containing a union of two other records, we would define:
+```
+namespace com.linkedin.pegasus.generator.examples
+
+record Question {
+  answerFormat: union[MultipleChoice, TextEntry]
+}
+```
+```
+namespace com.linkedin.pegasus.generator.examples
+
+record MultipleChoice {
+  answer: string
+}
+```
+
+```
+namespace com.linkedin.pegasus.generator.examples
+
+record TextEntry {
+  text: string
+}
+```
+
+Union with default value:
+```
+namespace com.linkedin.pegasus.generator.examples
+
+record Question {
+  answerFormat: union[MultipleChoice, TextEntry] = {"com.linkedin.pegasus.generator.examples.MultipleChoice" : {"answer" : "A"}}
+}
+```
+### Union with aliases
```

Comment:
1) Show that aliased union members can have doc strings and custom properties. Clarify that this is not supported for non-aliased union members.
2) Explain that either ALL union members must be aliased, or NONE at all. Having a "heterogeneously aliased union" is not allowed (the only exception is that the "null" member may be non-aliased, but you can leave this out because it seems like we plan on removing support for this).

Reply:
Updated.